samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -0,0 +1,166 @@
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.zkLeaderElectionListener = leaderElectionListener;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  public ZkLeaderElector(String processorIdStr, ZkUtils zkUtils) {
+    this.zkLeaderElectionListener = new ZkLeaderElectionListener();
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  // TODO: This should go away once we integrate with Zk based Job Coordinator
+  private String getHostName() {
+    try {
+      return InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      LOGGER.error("Failed to fetch hostname of the processor", e);
+      throw new SamzaException(e);
+    }
+  }
+
+  @Override
+  public boolean tryBecomeLeader() {
+    String currentPath = zkUtils.getEphemeralPath();
+
+    if (currentPath == null || currentPath.isEmpty()) {
+      currentPath = zkUtils.registerProcessorAndGetId(hostName);
+    }
+
+    List<String> children = zkUtils.getActiveProcessors();
+    LOGGER.debug(zLog("Current active processors - " + children));
+    int index = children.indexOf(ZkKeyBuilder.parseIdFromPath(currentPath));
+
+    if (children.size() == 0 || index == -1) {
+      throw new SamzaException("Looks like we are no longer connected to Zk. Need to reconnect!");
+    }
+
+    leaderId = ZkKeyBuilder.parseIdFromPath(children.get(0));
+    if (index == 0) {
+      LOGGER.info(zLog("Eligible to become the leader!"));
+      return true;
+    }
+
+    LOGGER.info("Index = " + index + " Not eligible to be a leader yet!");
+    String predecessor = children.get(index - 1);
+    if (!predecessor.equals(currentSubscription)) {
+      if (currentSubscription != null) {
+        LOGGER.debug(zLog("Unsubscribing data change for " + currentSubscription));
+        zkUtils.unsubscribeDataChanges(keyBuilder.getProcessorsPath() + "/" + currentSubscription, zkLeaderElectionListener);
+      }
+      currentSubscription = predecessor;
+      LOGGER.info(zLog("Subscribing data change for " + predecessor));
+      zkUtils.subscribeDataChanges(keyBuilder.getProcessorsPath() + "/" + currentSubscription, zkLeaderElectionListener);
+    }
+    /**
+     * Verify that the predecessor still exists. This step is needed because the ZkClient subscribes for data changes
+     * on the path, even if the path doesn't exist. Since we are using Ephemeral Sequential nodes, if the path doesn't
+     * exist during subscription, it is not going to get created in the future.
+     */
+    boolean predecessorExists = zkUtils.exists(keyBuilder.getProcessorsPath() + "/" + currentSubscription);
+    if (predecessorExists) {
+      LOGGER.info(zLog("Predecessor still exists. Current subscription is valid. Continuing as non-leader."));
+    } else {
+      try {
+        Thread.sleep(random.nextInt(1000));
```

Comment:
The point of the retry here is not because of the small window where the predecessor dies after subscription. It is done because the zkClient api subscribes to a path, even if it doesn't exist. let's say, you are p3 and you do a "getChildren", then you  see "p1, p2, p3, p4" -> Now before you do subscribe(p2), p2 goes away. However, the subscribe api will not fail if the path doesn't exists. So, we re-check with an "exists" api and make sure that subscription is actually on a valid active path. Otherwise, we subscribed to nothing and we will never climb up the ladder for leadership. does that make sense? I think I added a comment explaining why we are doing this.